x/ecocredit/server/core/bridge_receive.go

```diff
@@ -0,0 +1,135 @@
+package core
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	api "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+)
+
+// BridgeReceive bridges credits received from another chain.
+func (k Keeper) BridgeReceive(ctx context.Context, req *core.MsgBridgeReceive) (*core.MsgBridgeReceiveResponse, error) {
+	bridgeServiceAddr, err := sdk.AccAddressFromBech32(req.ServiceAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	class, err := k.stateStore.ClassTable().GetById(ctx, req.Project.ClassId)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("could not get class with id %s: %s", req.Project.ClassId, err.Error())
+	}
+
+	// first we check if there is an existing project
+	idx := api.ProjectClassKeyReferenceIdIndexKey{}.WithClassKeyReferenceId(class.Key, req.Project.ReferenceId)
+	it, err := k.stateStore.ProjectTable().List(ctx, idx)
+	if err != nil {
+		return nil, err
+	}
+
+	// we only want the first project that matches the reference ID, so we do not loop here.
+	var project *api.Project
+	if it.Next() {
+		var err error
+		project, err = it.Value()
+		if err != nil {
+			return nil, err
+		}
+	}
+	it.Close()
```

Comment:
talked with @clevinson and decided instead of erroring in the expectedly rare event that _multiple_ projects with the same Reference_Id exist within a credit class, we just pick the first project out of the iterator, if any.

This __does__ however leave one vulnerability, which i presume to be incredibly rare, but still worth mentioning. The scenario where things can go wrong:


- user creates Project with `VCS-001` as reference ID in `C01`
- user creates a batch from this project, using all TCO2 data that our bridge would use
- bridge service gets processes an event that __should__ bridge into `VCS-001` in that batch bob created, triggering a `MintBatchCredits` call
- this call would **fail** as `MintBatchCredits` is currently gated by the issuer address, which would be the user, not the bridge service address.

again, i presume this is not likely, but just want to be sure we are all aware and if its something we should be considering and if so if we should be preparing some sort of procedure if that scenario does occur.

Reply:
one remedy to eliminate this situation entirely: only match batches that have the same issuer address as the bridge service's address